compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEInsertDmaOutOfOrderBlock.cpp

```diff
@@ -0,0 +1,202 @@
+// Copyright 2025 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree-amd-aie/IR/AMDAIEOps.h"
+#include "iree-amd-aie/Transforms/Passes.h"
+#include "iree-amd-aie/Transforms/Utils/AMDAIEUtils.h"
+#define DEBUG_TYPE "iree-amdaie-insert-dma-out-of-order-block"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+namespace {
+
+/// Find the MM2S DMA start op corresponding to the given connection.
+FailureOr<AMDAIE::DMAStartOp> getMm2sDmaStart(
+    AMDAIE::ConnectionOp connectionOp,
+    DenseMap<AMDAIE::ChannelOp, SmallVector<AMDAIE::DMAStartOp>>
+        &mm2sChannelToDmaStarts) {
+  SmallVector<Value> srcChannels = connectionOp.getSourceChannels();
+  if (srcChannels.size() != 1) {
+    return connectionOp.emitOpError()
+           << "expected exactly one source channel for connection op";
+  }
+  auto channelOp = dyn_cast<AMDAIE::ChannelOp>(srcChannels[0].getDefiningOp());
+  if (!channelOp)
+    return connectionOp.emitOpError() << "expected an `amdaie.channel` op";
+  if (mm2sChannelToDmaStarts.count(channelOp) == 0) {
+    return connectionOp.emitOpError()
+           << "source channel does not have corresponding MM2S DMA start op";
+  }
+  if (mm2sChannelToDmaStarts[channelOp].size() != 1) {
+    return connectionOp.emitOpError()
+           << "expected exactly one MM2S DMA start op";
+  }
+  return mm2sChannelToDmaStarts[channelOp][0];
+}
+
+LogicalResult insertDmaOutOfOrder(
+    IRRewriter &rewriter, AMDAIE::ChannelOp s2mmChannelOp,
+    SmallVector<AMDAIE::DMAStartOp> &s2mmDmaStartOps,
+    DenseMap<AMDAIE::ChannelOp, SmallVector<AMDAIE::DMAStartOp>>
+        &mm2sChannelToDmaStarts) {
+  // If there is only one DMA start op, do nothing.
+  if (s2mmDmaStartOps.size() <= 1) return success();
+  // Merge the given DMA start ops into a single out-of-order block.
+  int8_t newRepeatCount = 0;
```

Comment:
I think repeat count can in principle go up to 256? I think it would be better to use an uint32_t and then check for overflow down below before creating the DMAStartOp.